### PR TITLE
Include py.typed file to make fastwarc PEP 561 compliant

### PR DESCRIPTION
**Pull request by jonded94**
Now that PR https://github.com/chatnoir-eu/chatnoir-resiliparse/pull/44 is in, to make this package *actually* work in a `mypy` context, we'd need to include a `py.typed` marker file.

Otherwise you'd see this error message:
```
error: Skipping analyzing "fastwarc.warc": module is installed, but missing library stubs or py.typed marker  [import-untyped]
```

Including this file makes this package properly PEP 561 compliant

### CODE REVIEW

**Comment by phoerious** [comment]
Did you check whether the file is actually included in the wheel? You might need to add it to the `package_data`. Also please rebase the branch to the latest develop.

**Comment by jonded94** [comment]
Yes, when I checked, it was both in the `sdist` and the actual wheels. The `.pyi` files are too btw.

I'll rebase this as soon as I opened another PR, I noticed that unfortunately a bunch of the type hints that were introduced are slightly wrong and/or missing.

**Comment by jonded94** [comment]
Hey @phoerious ,

unfortunately, adding and/or fixing wrong or missing type hints took longer than initially thought.

Patching a `py.typed` file into this package and testing what the output `mypy` was on the test files, specifically `tests/fastwarc/test_warc.py`, revealed that there were some issues with the current stub files.

The most central change of this PR is that places that accepted purely an `fastwarc.IOStream` before now accept both a `fastwarc.IOStream` or a generic `typing.BinaryIO`.

*Current state of type checking of tests:*

In general, I have to say that the state of the current tests is really not ideal to debug these kinds of things. It's practically entirely untyped code which creates a lot of noise for `mypy` which you have to sieve through to find the *actual* typing mishaps. This is the state of `mypy --strict tests/fastwarc/test_warc.py` on this branch, i.e. after fixing the issues in the stub files:
```
Found 112 errors in 1 file (checked 1 source file)
```

*Implementation changes that could be nice for the future:*

Another, relatively central issue is that `IOStream` does not implement a proper `typing.IO` interface. For this, following methods are missing/and or not correctly implemented:

```
IOStream.mode(self) -> str:
IOStream.name(self) -> str:
IOStream.closed(self) -> bool:
IOStream.fileno(self) -> int:
IOStream.isatty(self) -> bool:
IOStream.read(self, n: int = -1) -> bytes:  # needs default 'n = -1', meaning read everything
IOStream.readable(self) -> bool:
IOStream.readline(self, limit: int = -1) -> bytes:
IOStream.readlines(self, hint: int = -1) -> List[bytes]:
IOStream.seek(self, offset: int, whence: int = 0) -> int:  # needs 'whence'
IOStream.seekable(self) -> bool:
IOStream.truncate(self, size: int = None) -> int:
IOStream.writeable(self) -> bool:
IOStream.writelines(self, lines: list[bytes]) -> None:
```

Note that this is about implementing just a method interface, the actual non-abstract classes *can* raise a `io.UnsupportedOperation` for example, if that specific operation really can't be supported.

Also, for actually implementing `typing.MutableMapping`, the `WarcHeaderMap` would need:
```
WarcHeaderMap.__delitem__(self, key: str) -> None:
```

*TL;DR*

But nevertheless, I think this current state should already fix most issues that users have with this library in the wild. Let me know what you think about it.

**Comment by jonded94** [comment]
Slight correction: Since `gzip.GzipFile` for example does not properly implement `typing.BinaryIO` (it's missing `truncate()`), but it definitely *is* allowed as a possible type here, I went for some generic Readable/WriteableStream `Protocols`.

All python objects that implement necessary methods are now allowed in especially `ArchiveIterator` & `WarcRecord.write(...)`.

This maps nicely to what the Cython code actually does low-level, i.e. checking for methods `read`, `write`, `close`, ... on a generic Python object and then wrapping it with `PythonIOStreamAdapter`.

File: tests/fastwarc/test_warc.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import lz4.frame
 import io
 import os
+import hashlib
```
Comment (jonded94) [inline tests/fastwarc/test_warc.py:11]: For example for this test line (https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/774e33f45b5aef34fade03b465888925169e76b5/tests/fastwarc/test_warc.py#L330) the import was missing. How did this work previously?

**Comment by phoerious** [comment]
You can add the missing methods if you want, but they would all raise exceptions, unless you implement all the logic needed. I'm not sure whether this is actually useful. The IOStream classes are not drop-in replacements for their Python equivalents.

> the import was missing. How did this work previously?

I added explicit `__all__` in a recent commit. Previously, this was a hanger-on of the `from fastwarc.warc import *`. Just rebase the branch and everything should be fine.

**Comment by phoerious** [comment]
I'm still trying to figure out how to make the tokenless Codecov upload work on pull requests from forks. :thinking:

**Comment by jonded94** [comment]
>I'm not sure whether this is actually useful. The IOStream classes are not drop-in replacements for their Python equivalents.

I don't see that it's that clear that they couldn't do that. Properly implementing some centrally, well-known interfaces is always a plus in my book.

But this was rather about *removing* that the stub files wrongly assume that these interfaces were properly implemented, leading to `mypy` errors wherever you use these classes. This is now done and I don't see a strict need for anything further in the short-term future.

> I added explicit __all__ in a recent commit. Previously, this was a hanger-on of the from fastwarc.warc import *. Just rebase the branch and everything should be fine.

Ah, nice! Will do.

**Comment by phoerious** [comment]
I've updated the workflow file a bit. Please rebase again, so we can see whether the Codecov upload now works.

**Comment by jonded94** [comment]
Rebased on `d9f785ac40076260f9eb728a08e85d23b0d1d87f`, but still failed? :/

File: tests/fastwarc/test_warc.py (hunk 1 above)
Comment (phoerious) [reply tests/fastwarc/test_warc.py:11]: This change is unnecessary. There's already an `import hashlib` below the `import gzip` due to the rebase.

File: fastwarc/fastwarc/__init__.py [hunk 2]
```diff
@@ -15,3 +15,14 @@
...
+__all__ = [
+    "FileStream",
+    "GZipStream",
+    "LZ4Stream",
+    "FastWARCError",
+    "StreamError",
+    "ArchiveIterator",
+    "WarcRecord",
+    "WarcRecordType"
+]
```
Comment (phoerious) [inline fastwarc/fastwarc/__init__.py:28]: Since there aren't any unwanted global imports in this file, I didn't specify `__all_` explicitly. Is this needed for mypy? Otherwise I would suggest to remove it.

Comment (phoerious) [reply fastwarc/fastwarc/__init__.py:28]: Hmm, ok. It should be said that these are there only for legacy compatibility and explicit submodule imports should be preferred.

**Comment by jonded94** [comment]
Removed the unnecessary `hashlib` import.

File: fastwarc/fastwarc/stream_io.pyi [hunk 3]
```diff
@@ -36,28 +53,28 @@ class CompressingStream(IOStream):
...
     def __init__(
-        self, raw_stream: IOStream, quality: int = 11, lgwin: int = 22, lgblock: int = 0
+        self, raw_stream: Union[IOStream, BinaryIO, _ReadableStream, _WritableStream], quality: int = 11, lgwin: int = 22, lgblock: int = 0
     ) -> None: ...
 
 
 class GZipStream(CompressingStream):
     def __init__(
-        self, raw_stream: IOStream, compression_level: int = 9, zlib: bool = False
+        self, raw_stream: Union[IOStream, BinaryIO, _ReadableStream, _WritableStream], compression_level: int = 9, zlib: bool = False
```
Comment (phoerious) [inline fastwarc/fastwarc/stream_io.pyi:62]: One more question: Do we need `_ReadableStream` and `_WritableStream`? Isn't `IOStream` enough?

Comment (jonded94) [reply fastwarc/fastwarc/stream_io.pyi:62]: >We need at least close() and tell()

Thanks, I missed that!

I'm very happy to include `close` and `tell` as one of the required methods, I'll push a commit shortly.

In the case of `CompressingStream`s, `close` seems indeed to be called at `__dealloc__` while `tell` is called during `__cinit__`. There is one explicit call to `tell` in the `fastwarc.warc.ArchiveIterator` and `close` will probably called implicitly also through this `__dealloc__` method.

Possibly one should include a test for that, i.e. providing a minimal class that implements these, there are already tests that test with generic `io.BytesIO` and `open` objects.
 

>and at that point you are already at IOStream.

Do you mean `fastwarc.stream_io.IOStream` or a class that satisfies `typing.BinaryIO`? If `IOStream`: I would disagree with that, since the point of the `Protocol`s is to allow any arbitrary python class (as the Cython code does), i.e. *not* having to *inherit* from `IOStream`.

As a datapoint that possibly helps you understanding my standpoint: We're using `fastwarc` to a broad degree, but more for interacting with WARC files on already decompressed, custom streams. We are not using any of the stream classes of this package.

Comment (phoerious) [reply fastwarc/fastwarc/stream_io.pyi:62]: I was thinking of something like `Union[IOStream, BinaryIO]`. The former is satisfied by FastWARC's internal `stream_io` classes, the other is any other Python file-like object that would be wrapped in `PythonIOStreamAdapter`.

Comment (phoerious) [reply fastwarc/fastwarc/stream_io.pyi:62]: Think more about it, I would prefer `Union[IOStream, BinaryIO]`. It's stricter and therefore more future-safe.

Comment (phoerious) [reply fastwarc/fastwarc/stream_io.pyi:62]: One more thing: All members of `WarcRecordType` should be available as namespace-less flags, not juts `no_type` and `any_type`.

Comment (jonded94) [reply fastwarc/fastwarc/stream_io.pyi:62]: >If you want laxer duck typing, you could make IOStream a Protocol class via multi-inheritance.

This feels messy to me, IMHO.

>Thinking more about it, I would prefer Union[IOStream, BinaryIO]. It's stricter and therefore more future-safe.

Oh no, then maybe my point did not come across correctly. Doing it this way would break everything around our approach and not helping. 

Just take a look at `gzip.GzipFile` for example, just picking *something* from a battle-tested, well maintainted upstream library: https://github.com/python/cpython/blob/253c6a0b2f88b3327b7113860b99d665346fe43c/Lib/gzip.py#L142
This class *does* very much implement everything that's required for everything in `fastwarc`, but still *fails* a check against *typing.BinaryIO*. This is because it's missing `truncate()`, although it's nowhere used in this code.

Moreover, the entire point of `PythonIOStreamAdapter` is that it can wrap arbitrary classes, as long as they implement the minimal set of required methods. Forcing a `typing.BinaryIO` here would somewhat nullify that.

I'd *strongly* recommend leaving in the proper `Protocol` classes here, as they perfectly map on to the logic of this code and also would nullify the advantage that this PR has.

Comment (jonded94) [reply fastwarc/fastwarc/stream_io.pyi:62]: >One more thing: All members of WarcRecordType should be available as namespace-less flags, not juts no_type and any_type.

Sure, can do, but our of curiosity: What is the value of exposing every possible value of this enum as a separate variable at top-level? Is this something that has been done for backwards compatibility?

Comment (phoerious) [reply fastwarc/fastwarc/stream_io.pyi:62]: > This class does very much implement everything that's required for everything in fastwarc, but still fails a check against typing.BinaryIO. This is because it's missing truncate(), although it's nowhere used in this code.
> 
> Moreover, the entire point of PythonIOStreamAdapter is that it can wrap arbitrary classes, as long as they implement the minimal set of required methods. Forcing a typing.BinaryIO here would somewhat nullify that

Technically, the API documentation allows only "file-like" objects, so they should implement everything that's needed. But fair point. I'm fine with having a `Protocol` class then, but it shouldn't be split into readable and writable, since the interface doesn't guarantee that a read-only or write-only stream would be safe. You could technically pass a read-only stream to `ArchiveIterator`, but everything else must be read-write and should require all methods that `IOStream` has.

> Sure, can do, but our of curiosity: What is the value of exposing every possible value of this enum as a separate variable at top-level? Is this something that has been done for backwards compatibility?

Not really for compatibility, but for ease of use. `record.record_type == WarcRecordType.continuation` is a rather unwieldy expression and checking for record types is a pretty common action.

Comment (jonded94) [reply fastwarc/fastwarc/stream_io.pyi]: Went for a general `_GenericIOStream` protocol. 

Exposed all enum values at top level.

**Review by phoerious — APPROVED** [review]
